cluster_management/src/main/java/com/pinterest/rocksplicator/task/DedupTask.java

```diff
@@ -0,0 +1,113 @@
+  private final String taskCluster;
+  private final String job;
+  private final int adminPort;
+  private final String destStorePathPrefix;
+
+  public DedupTask(String srcStorePathPrefix, long resourceVersion, String partitionName,
+                   String taskCluster, String job, int adminPort,
+                   String destStorePathPrefix) {
+    this.srcStorePathPrefix = srcStorePathPrefix;
+    this.resourceVersion = resourceVersion;
+    this.partitionName = partitionName;
+    this.taskCluster = taskCluster;
+    this.job = job;
+    this.adminPort = adminPort;
+    this.destStorePathPrefix = destStorePathPrefix;
+  }
+
+  /**
+   * Execute the task.
+   * @return A {@link TaskResult} object indicating the status of the task and any additional
+   *         context information that can be interpreted by the specific {@link Task}
+   *         implementation.
+   */
+  @Override
+  public TaskResult run() {
+
+    if (srcStorePathPrefix.isEmpty() || resourceVersion == -1 || destStorePathPrefix.isEmpty()) {
+      String errMsg =
+          "Cancel the task, due to job command config map failed to provide all three of "
+              + "srcStorePathPrefix, resourceVersion, and destStorePathPrefix";
+      LOG.error(errMsg);
+      return new TaskResult(TaskResult.Status.CANCELED, errMsg);
+    }
+
+    // check task created with resource name and match prefix of target partition
+    String dbName = Utils.getDbName(partitionName);
+
+    try {
+
+      String srcStorePath =
+          String.format("%s/%s/%s", srcStorePathPrefix, String.valueOf(resourceVersion), dbName);
+
+      String destStorePath =
+          String.format("%s/%s/%s", destStorePathPrefix, String.valueOf(resourceVersion), dbName);
+
+      LOG.error(
+          String.format(
+              "DedupTask run to dedup partition: %s from source path: %s, to dest path: %s "
+                  + "Other info {cluster: %s, job: %s, resourceVersion: %d}", dbName,
+              srcStorePath, destStorePath, taskCluster, job, resourceVersion));
+
+      executeDedup(dbName, adminPort, srcStorePath, destStorePath);
+
+      LOG.error("DedupTask completed, with: success");
+      return new TaskResult(TaskResult.Status.COMPLETED, "DedupTask is completed!");
+    } catch (Exception e) {
+      LOG.error("Task dedup failed", e);
+      return new TaskResult(TaskResult.Status.FAILED, "DedupTask failed");
+    }
+
+  }
+
+  protected void executeDedup(String dbName, int adminPort, String srcStorePath,
+                              String destStorePath) throws RuntimeException {
+    try {
+      Utils.addDB(dbName, adminPort);
+      Utils.closeDB(dbName, adminPort);
+      Utils.restoreLocalDB(adminPort, dbName, srcStorePath, "127.0.0.1", adminPort);
+      LOG.error("restoreDB is done, begin compactDB");
+
+      Utils.compactDB(adminPort, dbName);
+      LOG.error("compactDB is done");
+
+      Utils.backupDB("127.0.0.1", adminPort, dbName, destStorePath);
```

Comment:
Have you considered splitting these into separate tasks (jobs)? My concern is that if compaction or backup fails, on retries you will unnecessarily repeat all the prior steps.

Reply:
good point. 
We can keep current implementation as a 1st version, since it won't repeat all the work even though it might repeat all function calls. For example, if 1st try failed at "upload", the task will repeat previous function calls, but, since local db still existed, download are incremental already with rocksdb api; so it will issue a restore call, but won't download anything. It might repeat a whole "compactDB", which could be an optimization we can do when we implement @cancel function in the Task.

Reply:
This does raise the concern, if any of these steps failed without cleaning local db (ie. current @cancel is empty), and task got re-distributed to other host to retry. Then, hosts are left with some zoombie dbs.. 
I can optimize this when I implement @cancel in Task

Reply:
This is a good point. 
Current implementation, with default TaskStateMode, interruption of a `running` task to `task_aborted', 'stopped', or 'task_error' , etc. will call the @cancel to clear local dbs if existed. Then, it will repeat all add/restore/compact/upload steps. 
Current implementation can prevent un-cleaned local dbs with potential repeated works, so, I will open another ticket to optimize this step.